packages/firestore/src/local/local_store.ts

```diff
@@ -242,24 +243,70 @@ export class LocalStore {
   }
   /* Accept locally generated Mutations and commit them to storage. */
   localWrite(mutations: Mutation[]): Promise<LocalWriteResult> {
+    const localWriteTime = Timestamp.now();
+    const keys = mutations.reduce(
+      (keys, m) => keys.add(m.key),
+      documentKeySet()
+    );
+
     return this.persistence.runTransaction(
       'Locally write mutations',
       'readwrite',
       txn => {
-        let batch: MutationBatch;
-        const localWriteTime = Timestamp.now();
-        return this.mutationQueue
-          .addMutationBatch(txn, localWriteTime, mutations)
-          .next(promisedBatch => {
-            batch = promisedBatch;
-            // TODO(koss): This is doing an N^2 update by replaying ALL the
-            // mutations on each document (instead of just the ones added) in
-            // this batch.
-            const keys = batch.keys();
-            return this.localDocuments.getDocuments(txn, keys);
-          })
-          .next((changedDocuments: MaybeDocumentMap) => {
-            return { batchId: batch.batchId, changes: changedDocuments };
+        // Load and apply all existing mutations. This lets us compute the
+        // current base state for all non-idempotent transforms before applying
+        // any additional user-provided writes.
+        return this.localDocuments
+          .getDocuments(txn, keys)
+          .next(existingDocs => {
+            // For non-idempotent mutations (such as `FieldValue.numericAdd()`),
+            // we record the base state in a separate patch mutation. This is
+            // later used to guarantee consistent values and prevents flicker
+            // even if the backend sends us an update that already includes our
+            // transform.
+            const baseMutations: Mutation[] = [];
+
+            for (const mutation of mutations) {
+              const maybeDoc = existingDocs.get(mutation.key);
+              if (!mutation.isIdempotent) {
+                // Theoretically, we should only include non-idempotent fields
+                // in this field mask as this mask is used to populate the base
+                // state for all DocumentTransforms.  By  including all fields,
+                // we incorrectly prevent rebasing of idempotent transforms
+                // (such as `arrayUnion()`) when any non-idempotent transforms
+                // are present.
+                const fieldMask = mutation.fieldMask;
+                if (fieldMask) {
+                  if (maybeDoc instanceof Document) {
+                    const baseValues = fieldMask.applyTo(maybeDoc.data);
```

Comment:
Can you port the latest Android? (Assign baseValues using ternary operator and include comment about Precondition.exists(true)).

Reply:
Done